services/graphql-server/directives/common/rest/rest.go

```diff
@@ -0,0 +1,265 @@
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/graphql-go/graphql"
+	"github.com/sirupsen/logrus"
+	"github.com/vektah/gqlparser/ast"
+)
+
+type RestParams struct {
+	url         string
+	method      string
+	contentType string
+	bodyArg     string
+	query       NameValueList
+	headers     NameValueList
+	timeoutMs   int
+}
+
+type httpClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+const (
+	defaultTimeoutMs = 10000
+	defaultMethod    = "GET"
+	defaultBodyArg   = "input"
+)
+
+var defaultContentType = contentTypes.JSON.name
+
+var RestMiddleware = middlewares.DirectiveDefinition{
+	MiddlewareFactory: func(serverCtx server.ServerContext, fieldDef *ast.FieldDefinition, directive *ast.Directive) middlewares.Middleware {
+		params := parseRestParams(directive)
+		client := createHTTPClient(params.timeoutMs)
+
+		return middlewares.ConcurrentLeaf(func(resolveParams graphql.ResolveParams) (interface{}, error) {
+			request, err := createHTTPRequest(serverCtx, params, resolveParams)
+			if err != nil {
+				return nil, err
+			}
+
+			return sendHTTPRequest(request, &client)
+		})
+	},
+}
+
+func parseRestParams(directive *ast.Directive) RestParams {
+	params := RestParams{}
+	var ok bool
+
+	args := directive.ArgumentMap(make(map[string]interface{}))
+
+	params.url, ok = args["url"].(string)
+	if !ok {
+		logrus.Panic("url argument missing from rest directive")
+	}
+
+	params.method = getMethod(args)
+	params.contentType = getContentTypeName(args)
+	params.timeoutMs = getTimeout(args)
+	params.bodyArg = getBodyArg(args)
+
+	params.query = getNameValueList(args, "query")
+	params.headers = getNameValueList(args, "headers")
+
+	return params
+}
+
+func getContentTypeName(args map[string]interface{}) string {
+	contentTypeName, ok := args["contentType"].(string)
+	if !ok {
+		return defaultContentType
+	}
+
+	contentType := getContentType(contentTypeName)
+	return contentType.name
+}
+
+func getTimeout(args map[string]interface{}) int {
+	timeoutInt64, ok := args["timeoutMs"].(int64)
+	if ok {
+		return int(timeoutInt64)
+	}
+
+	timeout, ok := args["timeoutMs"].(int)
+	if !ok {
+		return defaultTimeoutMs
+	}
+
+	return timeout
+}
+
+func getMethod(args map[string]interface{}) string {
+	method, ok := args["method"].(string)
+	if !ok {
+		return defaultMethod
+	}
+
+	return strings.ToUpper(method)
+}
+
+func getBodyArg(args map[string]interface{}) string {
+	bodyArg, ok := args["bodyArg"].(string)
+	if !ok {
+		return defaultBodyArg
+	}
+
+	return bodyArg
+}
+
+func getNameValueList(args map[string]interface{}, field string) NameValueList {
+	fieldArgs, ok := args[field].([]interface{})
+	if !ok {
+		return nil
+	}
+
+	nameValueList := make(NameValueList, 0, len(fieldArgs))
+
+	for _, arg := range fieldArgs {
+		nameValuePair := arg.(map[string]interface{})
+		nameValueList.append(nameValuePair["name"].(string), nameValuePair["value"].(string))
+	}
+
+	return nameValueList
+}
+
+func createHTTPClient(timeoutMs int) http.Client {
```

Comment:
I'm not sure, but I thought something like `7 * time.Millisecond` should just work, without needing to break it down to nanoseconds and back again

Reply:
`Duration` expects its input to be in nanoseconds, so I'm converting it from ms to ns, but I'm not converting it back at any point

Reply:
Yes, but `7 * time.Millisecond` is already a `time.Duration` by the magic of golang types

Reply:
That magic only works with number literals, but as much as I could find it doesn't work when the number you need comes from a variable (`timeoutMs`)

Reply:
https://play.golang.org/p/yqxfp4znTd0
Seems to work?